lib/dockerfile.ts

```diff
@@ -0,0 +1,223 @@
+						throw new DockerfileParseError('Non-string argument to WORKDIR');
+					}
+					workDir = entry.args;
+					break;
+				case 'RUN':
+					// Add this to the current set of commands
+					commands.push(Dockerfile.processRunArgs(entry.args));
+					break;
+			}
+		});
+
+		// Store any commands left over
+		if (commands.length > 0) {
+			this.actionGroups.push({
+				workDir,
+				commands,
+				fileDependencies: lastCopy,
+			});
+		}
+	}
+
+	// TODO: We can probably cache this value
+	/**
+	 * Get a list of action groups which should be re-executed if any of
+	 * the files passed to this function have changed.
+	 *
+	 * @param files A list of files whose contents have changed. The path
+	 * should be relative to the build context
+	 */
+	public getActionGroupsFromChangedFiles(files: string[]): DockerfileActionGroup[] {
+		let actionGroupIdx = 0;
+		for (const actionGroup of this.actionGroups) {
+			// FIXME: COPY src/ src/ won't work due to the fact that there isn't a glob on the end of
+			// the target - we can fix this by silenty converting src/ to src/*
+
+			// For every dependency, see if it matches the current action group
+			const matches = _.uniq(_.reduce(
+				actionGroup.fileDependencies,
+				(matchList, { localPath }) => {
+					const matchedFiles = minimatch.match(files, localPath);
+					return matchList.concat(matchedFiles);
+				},
+				[] as string[],
+			));
+
+			// If any of the files which have changed we need to return all
+			// action groups which follow, as they could depend on either
+			// one of the trigger files, or the output of a previous action
+			// group
+			if (matches.length > 0) {
+				return this.actionGroups.slice(actionGroupIdx);
+			}
+
+			++actionGroupIdx;
+		}
+
+		return [];
+	}
+
+	private static copyArgsToFileDeps(workDir: string, copyArgs: string[]): FileDependency[] {
+
+		// Remove any flags
+		copyArgs = Dockerfile.removeDashedArgs(copyArgs);
+
+		// The last entry into the array should be the destination
+		let maybeDest = copyArgs.pop();
+
+		if (maybeDest == null) {
+			throw new DockerfileParseError('Incorrect argument count for COPY');
+		}
+
+		let dest: string = maybeDest;
+
+		// If the destination is already absolute, the workdir doesn't come into it,
+		// but a relative destination means relative to the working directory
+		if (!path.isAbsolute(dest)) {
+			dest = path.join(workDir, dest);
+		}
+
+		// TODO: Do we need to handle a \ for windows?
+		const isDir = dest.endsWith('/');
+
+		return copyArgs.map((arg) => {
+
+			if (isDir) {
+				dest = path.join(dest, arg);
+			}
+
+			return {
+				localPath: arg,
+				containerPath: dest,
+			};
+		});
+	}
+
+	private static processRunArgs(command: string | string[]): string {
+		if (_.isString(command)) {
+			return command;
+		}
+		return command.join(' ');
+	}
+
+	private static removeDashedArgs(command: string[]): string[] {
+		let wasDashedArg = false;
+		return _.reduce(command, (args, arg) => {
```

Comment:
I think this can be:
```typescript
return _.reject(command, (arg) => {
	if (wasDashedArg) {
		wasDashedArg = false
		return true
	}
	if (arg.startsWith('--')) {
		wasDashedArg = true
	}
	return wasDashedArg
})